### PR TITLE
feature: support  XA transaction mode

### PR DESCRIPTION
**Pull request by sharajava**
### Ⅰ. Describe what this PR did

Provide XA mode to Seata. Issue #2022 

### Ⅳ. Describe how to verify it

all test cases can be found in XAModeTest2

### CODE REVIEW

**Comment by lovepoem** [comment]
Good job

**Comment by a364176773** [comment]
exciting

File: test/src/test/java/io/seata/xa/XAModeTest2.java [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+
+public class XAModeTest2 {
+
+    private static final int testRecordId = 888;
+    private static final String testRecordName = "xxx";
+    private static final long testTid = 1582688600006L;
+    private static final String mockXid = "127.0.0.1:8091:" + testTid;
+    private static final long mockBranchId = testTid + 1;
+
+    private static final String pg_jdbcUrl = "jdbc:postgresql://127.0.0.1:5432/postgres";
```
Comment (slievrly) [inline test/src/test/java/io/seata/xa/XAModeTest2.java]: Unable to connect to 127.0.0.1 database, CI occurred a large number of error logs. Please use mock or disable, and remove all printf logs.

**Comment by sharajava** [comment]
Samples For XA mode: https://github.com/seata/seata-samples/pull/324

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                                       String applicationData) throws TransactionException {
+        return finishBranch(false, branchType, xid, branchId, resourceId, applicationData);
+    }
+
+    private BranchStatus finishBranch(boolean committed, BranchType branchType, String xid, long branchId, String resourceId,
+                                       String applicationData) throws TransactionException {
+        XAXid xaBranchXid = XAXidBuilder.build(xid, branchId);
+        Resource resource = dataSourceCache.get(resourceId);
+        if (resource instanceof AbstractDataSourceProxyXA) {
+            try (ConnectionProxyXA connectionProxyXA = ((AbstractDataSourceProxyXA)resource).getConnectionForXAFinish(xaBranchXid)) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java:66]: clientA, clientB->TC      clientA branchCommit  but clientA ->TC  channel is bad and XAConn is ok, so TC send request to clientB.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    connectionProxyXA.xaCommit(xid, branchId, applicationData);
+                    LOGGER.info(xaBranchXid + " was committed.");
+                    return BranchStatus.PhaseTwo_Committed;
+                } else {
+                    connectionProxyXA.xaRollback(xid, branchId, applicationData);
+                    LOGGER.info(xaBranchXid + " was rolled back.");
+                    return BranchStatus.PhaseTwo_Rollbacked;
+                }
+            } catch (XAException xe) {
+                throw new TransactionException("XA Exception " + xe.getMessage(), xe);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java:77]: throw exception will result in encode error.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                    // FIXME: case of PhaseTwo_CommitFailed_Unretryable
+                    return BranchStatus.PhaseTwo_CommitFailed_Retryable;
+                } else {
+                    LOGGER.info(xaBranchXid + " rollback failed since " + sqle.getMessage(), sqle);
+                    // FIXME: case of PhaseTwo_RollbackFailed_Unretryable
+                    return BranchStatus.PhaseTwo_RollbackFailed_Retryable;
+                }
+            }
+        } else {
+            throw new TransactionException("Unknown Resource for XA resource " + resourceId + " " + resource);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java:90]: throw exception will result in encode error.

**Comment by slievrly** [comment]
https://github.com/seata/seata-samples/pull/324#issue-392770694
case check
- [ ] http://localhost:8084/purchase  committed
- [ ] http://localhost:8084/purchase?rollback=true  rollbacked
- [ ] http://localhost:8084/purchase , accountXa service down , rollbacked

File: rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    @Override
+    public PrintWriter getLogWriter() throws SQLException {
+        dataSourceCheck();
+        return dataSource.getLogWriter();
+    }
+
+    @Override
+    public void setLogWriter(PrintWriter out) throws SQLException {
+        dataSourceCheck();
+        if (dataSource == null) {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java:131]: Oh, useless code. Just removed.

File: rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+    @Override
+    public boolean isWrapperFor(Class<?> iface) throws SQLException {
+        return iface != null && iface.isInstance(this);
+
+    }
+
+    protected void dataSourceCheck() {
+        if (dataSource == null) {
+            throw new UnsupportedOperationException();
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java:118]: ok

File: rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        if (value.isKept()) {
+            T x = keeper.get(key);
+            if (x != value) {
+                throw new ShouldNeverHappenException("something wrong with keeper, keeping[" + x +
+                    "] but[" + value + "] is also kept with the same key[" + key + "]");
+            }
+            return value;
+        }
+        T x = keeper.put(key, value);
+        if (x != null) {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java:166]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ConnectionProxyXA.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+            } catch (TransactionException te) {
+                cleanXABranchContext();
+                throw new SQLException("failed to register xa branch " + xid + " since " + te.getCode() + ":" + te.getMessage(), te);
+            }
+            // 2. build XA-Xid with xid and branchId
+            this.xaBranchXid = XAXidBuilder.build(xid, branchId);
+            try {
+                // 3. XA Start
+                xaResource.start(this.xaBranchXid, XAResource.TMNOFLAGS);
+            } catch (XAException e) {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/ConnectionProxyXA.java:144]: Prefer to show different message on different cause of SQLException.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java (hunk 4 above)
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java:90]: In fact, this is a case that SHOULD NEVER happen.

Just merged XAException and  SQLException into retryable status.

File: rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+    protected String resourceGroupId;
+
+    protected BranchType branchType;
+
+    protected String dbType;
+
+    protected Driver driver;
+
+    private ConcurrentHashMap<String, T> keeper = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java:51]: As a private member variable, declaring an interface instead here seems not necessary.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/AbstractDataSourceProxyXA.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    public void forceClosePhysicalConnection(XAXid xaXid) throws SQLException {
+        ConnectionProxyXA connectionProxyXA = lookup(xaXid.toString());
+        if (connectionProxyXA != null) {
+            connectionProxyXA.close();
+            Connection physicalConn = connectionProxyXA.getWrappedConnection();
+            if (physicalConn instanceof PooledConnection) {
+                physicalConn = ((PooledConnection)physicalConn).getConnection();
+            }
+            // Force close the physical connection
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/AbstractDataSourceProxyXA.java:59]: Physical connection is wrapped in the PooledConnection.

File: rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    @Override
+    public Logger getParentLogger() throws SQLFeatureNotSupportedException {
+        dataSourceCheck();
+        return dataSource.getParentLogger();
+    }
+
+    @Override
+    public T keep(String key, T value) {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/BaseDataSourceResource.java:156]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/AbstractDataSourceProxyXA.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+public abstract class AbstractDataSourceProxyXA extends BaseDataSourceResource<ConnectionProxyXA> {
+
+    protected static final String DEFAULT_RESOURCE_GROUP_ID = "DEFAULT_XA";
+
+    /**
+     * Get a ConnetionProxyXA instance for finishing XA branch(XA commit/XA rollback)
+     * @return ConnetionProxyXA instance
+     * @throws SQLException exception
+     */
+    public ConnectionProxyXA getConnectionForXAFinish(XAXid xaXid) throws SQLException {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/AbstractDataSourceProxyXA.java:38]: Just keep it until a better name found.

File: rm-datasource/src/main/java/io/seata/rm/datasource/util/XAUtils.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        throw new SQLException("xa not support dbType: " + dbType);
+    }
+
+    private static XAConnection createOracleXAConnection(Connection physicalConnection, String xaConnectionClassName) throws XAException, SQLException {
+        try {
+            Class xaConnectionClass = Class.forName(xaConnectionClassName);
+            Constructor<XAConnection> constructor = xaConnectionClass.getConstructor(Connection.class);
+            return constructor.newInstance(physicalConnection);
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/util/XAUtils.java:77]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/util/XAUtils.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+                // return OracleUtils.OracleXAConnection(physicalConn);
+                String physicalConnClassName = physicalConn.getClass().getName();
+                if ("oracle.jdbc.driver.T4CConnection".equals(physicalConnClassName)) {
+                    return createOracleXAConnection(physicalConn, "oracle.jdbc.driver.T4CXAConnection");
+                } else {
+                    return createOracleXAConnection(physicalConn, "oracle.jdbc.xa.client.OracleXAConnection");
+                }
+            } catch (XAException xae) {
+                LOGGER.error("create xaConnection error", xae);
+                return null;
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/util/XAUtils.java:58]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/util/JdbcUtils.java [hunk 15]
```diff
@@ -45,4 +49,89 @@ static DbTypeParser getDbTypeParser() {
...
+        XAConnection xaConnection = null;
+        try {
+            xaConnection = dataSource.getXAConnection();
+            try (Connection connection = xaConnection.getConnection()) {
+                String jdbcUrl = connection.getMetaData().getURL();
+                dataSourceResource.setResourceId(buildResourceId(jdbcUrl));
+                String driverClassName = com.alibaba.druid.util.JdbcUtils.getDriverClassName(jdbcUrl);
+                dataSourceResource.setDriver(loadDriver(driverClassName));
+                dataSourceResource.setDbType(com.alibaba.druid.util.JdbcUtils.getDbType(jdbcUrl, driverClassName));
+            } finally {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/util/JdbcUtils.java:92]: Just refactor it in another way.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java (hunk 3 above)
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java:77]: Merged SQLException and XAException cases.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ExecuteTemplateXA.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+                } else {
+                    throw new SQLException(ex);
+                }
+
+            }
+            if (autoCommitStatus) {
+                try {
+                    // XA End & Prepare
+                    connectionProxyXA.commit();
+                } catch (Throwable ex) {
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/ExecuteTemplateXA.java:74]: Cases that need NOT to rollback is marked as a  specific SQLState in the SQLException thrown by commit().

File: rm-datasource/pom.xml [hunk 17]
```diff
@@ -109,5 +104,16 @@
...
+        <dependency>
+            <groupId>com.alibaba</groupId>
+            <artifactId>druid</artifactId>
+            <scope>provided</scope>
+            <optional>true</optional>
+        </dependency>
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/pom.xml:116]: Just depended when testing NOT runtime.

File: rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java (hunk 2 above)
Comment (sharajava) [reply rm-datasource/src/main/java/io/seata/rm/datasource/xa/ResourceManagerXA.java:66]: Yes, such cases should be considered in enhancement later.

Just keep the basic logic by now.

**Pull request merged**